### PR TITLE
[Spark] Support ALTER TABLE CLUSTER BY

### LINKED ISSUE #2468: [Feature Request] Support ALTER TABLE CLUSTER BY
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Overview

Uber issue: https://github.com/delta-io/delta/issues/1874

This issue tracks support for ALTER TABLE CLUSTER BY to change clustering columns on a clustered table.

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [x] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [ ] No. I cannot contribute this feature at this time.

### PR DESCRIPTION
**Pull request by zedtang**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Resolves #2468.

This PR adds support for ALTER TABLE CLUSTER BY syntax for clustered tables:
* `ALTER TABLE CLUSTER BY (col1, col2, ...)` to change the clustering columns
* `ALTER TABLE CLUSTER BY NONE` to remove the clustering columns

## How was this patch tested?

Add new unit tests.

## Does this PR introduce _any_ user-facing changes?

Yes, it introduces ALTER TABLE CLUSTER BY syntax.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala [hunk 1]
```diff
@@ -193,6 +193,24 @@ trait ClusteredTableUtilsBase extends DeltaLogging {
...
+  def getClusteringDomainMetadataForAlterTableClusterBy(
+      newLogicalClusteringColumns: Seq[String],
+      txn: OptimisticTransaction): Seq[DomainMetadata] = {
+    val newClusteringColumns =
+      newLogicalClusteringColumns.map(ClusteringColumn(txn.metadata.schema, _))
+    val clusteringMetadataDomainOpt =
+      if (txn.snapshot.domainMetadata.exists(_.domain == ClusteringMetadataDomain.domainName)) {
+        Some(ClusteringMetadataDomain.fromClusteringColumns(newClusteringColumns).toDomainMetadata)
+      } else {
+        None
```
Comment (dabao521) [inline spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala:209]: For clustered table the DomainMetadata should always be there. If so we don't need to handle `None`

**Review by dabao521 — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala (hunk 1 above)
Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala:209]: makes sense, updated

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 2]
```diff
@@ -755,6 +757,18 @@ class DeltaCatalog extends DelegatingCatalogExtension
...
           table, featureName, truncateHistory = truncateHistory).run(spark)
 
+      case (t, clusterBy) if t == classOf[ClusterBy] =>
+        clusterBy.asInstanceOf[Seq[ClusterBy]].foreach { c =>
+          if (c.clusteringColumns.nonEmpty) {
+            val clusterBySpec = ClusterBySpec(c.clusteringColumns.toSeq)
+            validateClusterBySpec(Some(clusterBySpec), table.schema())
+          }
+          if (!ClusteredTableUtils.isSupported(table.initialSnapshot.protocol)) {
+            throw DeltaErrors.alterClusterByNotOnLiquidDeltaTableException()
```
Comment (weiluo-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:767]: Add a test for this?

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:767]: done

**Review by imback82 — APPROVED** [review]

File: spark/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4 [hunk 3]
```diff
@@ -88,6 +88,8 @@ statement
         DROP CONSTRAINT (IF EXISTS)? name=identifier                    #dropTableConstraint
     | ALTER TABLE table=qualifiedName
         DROP FEATURE featureName=featureNameValue (TRUNCATE HISTORY)?   #alterTableDropFeature
+    | ALTER TABLE table=qualifiedName
+        (clusterBySpec | CLUSTER BY NONE)                               #alterTableClusterBy
```
Comment (imback82) [inline spark/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:92]: FYI, we will add this to OSS Spark but this is a stop gap before the next Spark release.

**Review by weiluo-db — APPROVED** [review]

**Pull request closed without merging**